### PR TITLE
Add AsyncHandler

### PR DESCRIPTION
**Pull request by xerial**
#2

### CODE REVIEW

**Comment by xerial** [comment]
Here is a quick experimental result. Performed logging of 100,000 records with 4 core MacBook Pro (Intel Core i7). To simulate heavy logging process, the log handler ran `Thread.sleep(1)` (just waiting 1 msec.). The log messages was discarded without using any memory or file.  I compared the performance of wvlet-log's default handler, AsyncHandler (added in this PR), twitter-util logging handler, and its QueueingHandler for asynchronous processing. 

For each handler, the test code generated 100 log records using 4 concurrent threads and this process will be repeated 100 times. 
wvlet_log_performance_-_google_sheets

This result does not mean synchronous logging is slow, because the log handler needs to wait 1 msec for every log request, and also it does not include the background processing time of the asynchronous log handlers. The result only shows the asynchronous logging is useful to avoid the delay of heavy logging process (e.g., writing to a file, network, etc.)

However if I used NullHandler, which outputs nothing, async logging always had a higher overhead compared to synchronous logging methods. This indicates we should use AsyncHandler only if we know the overhead of log handler is large.

Another finding is com.twitter.util.logging.QueueingHandler often drops log records under high pressure (e.g., just writing 1,000 log records in parallel). It showed many warning: `log queue overflow`.

File: src/main/scala/wvlet/log/LogRecord.scala [hunk 1]
```diff
@@ -28,21 +30,26 @@ case class LogSource(path: String, fileName: String, line: Int, col: Int) {
...
+  def apply(level: LogLevel, source: LogSource, message: String): LogRecord = {
     LogRecord(level, Some(source), message, None)
   }
 
-  def apply(level:LogLevel, source:LogSource, message:String, cause:Throwable) : LogRecord = {
+  def apply(level: LogLevel, source: LogSource, message: String, cause: Throwable): LogRecord = {
     LogRecord(level, Some(source), message, Some(cause))
   }
+
+  private[log] val leafLoggerNameCache = new mutable.WeakHashMap[String, String]
```
Comment (xerial) [inline src/main/scala/wvlet/log/LogRecord.scala:47]: This is for performance improvement

**Comment by xerial** [comment]
@Lewuathe When you have a time, please review this PR.

**Comment by xerial** [comment]
cc: @okapies

**Comment by xerial** [comment]
This chart shows the overhead of async logging:
wvlet_log_performance_-_google_sheets

**Comment by xerial** [comment]
@Lewuathe PTAL

File: README.md [hunk 2]
```diff
@@ -141,6 +146,24 @@ logger.resetHandler(new LogRotationHandler(
...
 
+### Asynchronous logging
+
+If you know your LogHandler is a heavy process (e.g., writing to network or slow disks), you can use 
+`AsyncHandler` to do the logging in a background thread:
+
+```scala
+val asyncHandler = new AsyncHandler(a heavy parent log handler)
+try {
+  logger.resetHandler(async)
```
Comment (Lewuathe) [inline README.md]: s/async/asyncHandler/

File: src/main/scala/wvlet/log/AsyncHandler.scala [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        }
+      }
+    }
+  })
+
+  override def flush(): Unit = {
+    val records = Seq.newBuilder[jl.LogRecord]
+    guard {
+      while (!queue.isEmpty) {
+        records += queue.pollFirst()
```
Comment (Lewuathe) [inline src/main/scala/wvlet/log/AsyncHandler.scala:55]: Can we write simply like this?

```
while (!queue.isEmpty) {
  parent.publish(queue.pollFirst())
}
```

File: src/main/scala/wvlet/log/io/StopWatch.scala [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+      (elapsedTimeAccumulated + diff) / NANO_UNIT
+    }
+    else {
+      elapsedTimeAccumulated / NANO_UNIT
+    }
+  }
+
+  /**
+    * Reset the stop watch. The subsequent calls to
+    * getElapsedTime or getIntervalTiem will measure the time interval
```
Comment (Lewuathe) [inline src/main/scala/wvlet/log/io/StopWatch.scala]: s/getIntervalTiem/getIntervalTime/

File: src/main/scala/wvlet/log/io/StopWatch.scala [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+  *
+  * When measuring Scala (Java) code performances, you should take the average of execution times and reorder
+  * the code block execution orders, because JVM has JIT compiler, which optimizes the code at runtime.
+  * And also cache usage and the running state of the garbage core (GC) affects
+  * the code performance. By repeating the executions of the entire or individual blocks with the `repeat` option,
+  * you can avoid such pitfalls of benchmarking.
+  *
+  * @author leo
+  */
+trait Timer extends Serializable {
```
Comment (Lewuathe) [inline src/main/scala/wvlet/log/io/StopWatch.scala:90]: `Timer` trait looks very useful 😄

File: src/main/scala/wvlet/log/io/StopWatch.scala [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+  def stop: Double = {
+    if (state == State.STOPPED) {
+      return 0.0
+    }
+
+    // elapsed time
+    val now = System.nanoTime()
+    val diff = now - lastSystemTime
+    elapsedTimeAccumulated += diff
+    lastSystemTime = now;
```
Comment (Lewuathe) [inline src/main/scala/wvlet/log/io/StopWatch.scala]: Semicolon is not needed.

File: src/main/scala/wvlet/log/io/StopWatch.scala [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+
+    state = State.STOPPED;
```
Comment (Lewuathe) [inline src/main/scala/wvlet/log/io/StopWatch.scala]: ditto

File: src/main/scala/wvlet/log/io/StopWatch.scala [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+  /**
+    * Resume the timer
+    */
+  def resume {
+    if (state == State.RUNNING) {
+      return
+    }
+
+    lastSystemTime = System.nanoTime();
+    state = State.RUNNING;
```
Comment (Lewuathe) [inline src/main/scala/wvlet/log/io/StopWatch.scala]: ditto

File: src/main/scala/wvlet/log/io/StopWatch.scala [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+
+  private[StopWatch] object State extends Enumeration {
+    val RUNNING, STOPPED = Value
+  }
+
+  private var lastSystemTime        : Double = System.nanoTime
+  private var elapsedTimeAccumulated: Double = 0L
+  private var state                          = State.RUNNING
+
+  private val NANO_UNIT: Double = 1000000000L
```
Comment (Lewuathe) [inline src/main/scala/wvlet/log/io/StopWatch.scala]: `NANO_UNIT` can be `1000000000d` instead of `1000000000L`?

**Review by Lewuathe — APPROVED** [review]
Though some minor comments, LGTM!

File: src/main/scala/wvlet/log/AsyncHandler.scala (hunk 3 above)
Comment (xerial) [reply src/main/scala/wvlet/log/AsyncHandler.scala:55]: Doing publish in the guard scope will increase the blocking time of the logger. We should quickly extract remaining log messages.

File: src/main/scala/wvlet/log/io/StopWatch.scala (hunk 9 above)
Comment (xerial) [reply src/main/scala/wvlet/log/io/StopWatch.scala]: Yes. I'll fix it

File: src/main/scala/wvlet/log/io/StopWatch.scala (hunk 4 above)
Comment (xerial) [reply src/main/scala/wvlet/log/io/StopWatch.scala]: 👍

File: src/main/scala/wvlet/log/io/StopWatch.scala (hunk 6 above)
Comment (xerial) [reply src/main/scala/wvlet/log/io/StopWatch.scala]: Actually I copied this code from Java 4 years ago... lol

**Pull request merged**